images/sds-lvm-csi/pkg/utils/func.go

```diff
@@ -157,25 +215,92 @@ func GetVGName(ctx context.Context, kc client.Client, lvmVG map[string]string, n
 		ListMeta: metav1.ListMeta{},
 		Items:    []v1alpha1.LvmVolumeGroup{},
 	}
-	err = kc.List(ctx, listLvgs)
+
+	for attempt := 0; attempt < KubernetesApiRequestLimit; attempt++ {
+		err = kc.List(ctx, listLvgs)
+		if err == nil {
+			break
+		}
+		time.Sleep(KubernetesApiRequestTimeout)
+	}
 	if err != nil {
-		return "", "", err
+		return "", "", fmt.Errorf("after %d attempts of getting LvmVolumeGroups, last error: %w", KubernetesApiRequestLimit, err)
 	}
 
 	for _, lvg := range listLvgs.Items {
-		_, ok := lvmVG[lvg.Spec.ActualVGNameOnTheNode]
-		if ok && lvg.Status.Nodes[0].Name == nodeName {
-			if LvmType == LLVTypeThin {
-				for _, thinPool := range lvg.Status.ThinPools {
-					for _, tp := range lvmVG {
-						if thinPool.Name == tp {
-							return lvg.Name, lvg.Spec.ActualVGNameOnTheNode, nil
+
+		log.Trace(fmt.Sprintf("[GetLVMVolumeGroupParams] process lvg: %+v", lvg))
+
+		_, ok := lvmVG[lvg.Name]
+		if ok {
+			log.Info(fmt.Sprintf("[GetLVMVolumeGroupParams] found lvg from storage class: %s", lvg.Name))
+			log.Info(fmt.Sprintf("[GetLVMVolumeGroupParams] lvg.Status.Nodes[0].Name: %s, prefferedNode: %s", lvg.Status.Nodes[0].Name, nodeName))
+			if lvg.Status.Nodes[0].Name == nodeName {
+				if LvmType == LLVTypeThin {
+					for _, thinPool := range lvg.Status.ThinPools {
+						for _, tp := range lvmVG {
+							if thinPool.Name == tp {
+								return lvg.Name, lvg.Spec.ActualVGNameOnTheNode, nil
+							}
 						}
 					}
 				}
+				return lvg.Name, lvg.Spec.ActualVGNameOnTheNode, nil
 			}
-			return lvg.Name, lvg.Spec.ActualVGNameOnTheNode, nil
+		} else {
+			log.Info(fmt.Sprintf("[GetLVMVolumeGroupParams] skip lvg: %s", lvg.Name))
 		}
+
 	}
 	return "", "", errors.New("there are no matches")
 }
+
+func GetLVMVolumeGroup(ctx context.Context, kc client.Client, lvgName, namespace string) (*v1alpha1.LvmVolumeGroup, error) {
+	var lvg v1alpha1.LvmVolumeGroup
+	var err error
+
+	for attempt := 0; attempt < KubernetesApiRequestLimit; attempt++ {
+		err = kc.Get(ctx, client.ObjectKey{
+			Name:      lvgName,
+			Namespace: namespace,
+		}, &lvg)
+
+		if err == nil {
+			return &lvg, nil
+		}
+		time.Sleep(KubernetesApiRequestTimeout)
+	}
+	return nil, fmt.Errorf("after %d attempts of getting LvmVolumeGroup %s in namespace %s, last error: %w", KubernetesApiRequestLimit, lvgName, namespace, err)
+}
+
+func GetLVMVolumeGroupCapacity(lvg v1alpha1.LvmVolumeGroup) (vgCapacity resource.Quantity, err error) {
```

Comment:
How do you think if we change the name to GetLVMVolumeGroupFreeSpace/SpaceLeft as the Capacity in the end means more about total space, not left one.